File: frontend/lib/sanitize.ts
/**
 * Input sanitization utilities for user-provided text fields.
 * Prevents XSS, injection attacks, and ensures data quality.
 */

const MAX_NAME_LENGTH = 100;

/**
 * Sanitizes a name field (department, squad, etc.)
 * - Trims leading/trailing whitespace
 * - Collapses multiple spaces into one
 * - Removes HTML tags and script content
 * - Removes control characters
 * - Limits length
 */
export const sanitizeName = (
  input: string,
  maxLength: number = MAX_NAME_LENGTH,
): string => {
  if (!input) return "";

  return (
    input
      // Remove HTML tags
      .replace(/<[^>]*>/g, "")
      // Remove script-like patterns
      .replace(/javascript:/gi, "")
      .replace(/on\w+=/gi, "")
      // Remove control characters (except newlines and tabs which get collapsed)
      .replace(/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/g, "")
      // Collapse multiple whitespace to single space
      .replace(/\s+/g, " ")
      // Trim whitespace
      .trim()
      // Limit length
      .slice(0, maxLength)
  );
};

/**
 * Validates that a sanitized name is acceptable
 * Returns an error message if invalid, or null if valid
 */
export const validateName = (
  name: string,
  fieldName: string = "Name",
  minLength: number = 1,
  maxLength: number = MAX_NAME_LENGTH,
): string | null => {
  if (!name || name.length < minLength) {
    return `${fieldName} must be at least ${minLength} character${minLength > 1 ? "s" : ""}`;
  }

  if (name.length > maxLength) {
    return `${fieldName} must be less than ${maxLength} characters`;
  }

  // Check for only whitespace/special chars
  if (!/[a-zA-Z0-9]/.test(name)) {
    return `${fieldName} must contain at least one letter or number`;
  }

  return null;
};

/**
 * Combined sanitize and validate - returns { value, error }
 */
export const sanitizeAndValidateName = (
  input: string,
  fieldName: string = "Name",
  minLength: number = 1,
  maxLength: number = MAX_NAME_LENGTH,
): { value: string; error: string | null } => {
  const value = sanitizeName(input, maxLength);
  const error = validateName(value, fieldName, minLength, maxLength);
  return { value, error };
};

File: backend/internal/sanitize/sanitize.go
// Package sanitize provides input sanitization utilities for user-provided text.
package sanitize

import (
	"regexp"
	"strings"
	"unicode"
)

const MaxNameLength = 100

var (
	// Patterns to remove
	htmlTagPattern    = regexp.MustCompile(`<[^>]*>`)
	jsProtocolPattern = regexp.MustCompile(`(?i)javascript:`)
	eventHandlerPattern = regexp.MustCompile(`(?i)on\w+=`)
	multiSpacePattern = regexp.MustCompile(`\s+`)
)

// Name sanitizes a name field (department, squad, etc.)
// - Trims leading/trailing whitespace
// - Collapses multiple spaces into one
// - Removes HTML tags and script content
// - Removes control characters
// - Limits length
func Name(input string, maxLength int) string {
	if input == "" {
		return ""
	}

	if maxLength <= 0 {
		maxLength = MaxNameLength
	}

	// Remove HTML tags
	result := htmlTagPattern.ReplaceAllString(input, "")

	// Remove javascript: protocol
	result = jsProtocolPattern.ReplaceAllString(result, "")

	// Remove event handlers
	result = eventHandlerPattern.ReplaceAllString(result, "")

	// Remove control characters (except standard whitespace)
	result = strings.Map(func(r rune) rune {
		if unicode.IsControl(r) && r != ' ' && r != '\t' && r != '\n' && r != '\r' {
			return -1
		}
		return r
	}, result)

	// Collapse multiple whitespace to single space
	result = multiSpacePattern.ReplaceAllString(result, " ")

	// Trim whitespace
	result = strings.TrimSpace(result)

	// Limit length
	if len(result) > maxLength {
		result = result[:maxLength]
	}

	return result
}

// ValidateName checks if a sanitized name is acceptable.
// Returns an error message if invalid, or empty string if valid.
func ValidateName(name string, fieldName string, minLength int, maxLength int) string {
	if minLength <= 0 {
		minLength = 1
	}
	if maxLength <= 0 {
		maxLength = MaxNameLength
	}

	if len(name) < minLength {
		if minLength == 1 {
			return fieldName + " is required"
		}
		return fieldName + " must be at least " + string(rune(minLength+'0')) + " characters"
	}

	if len(name) > maxLength {
		return fieldName + " must be less than " + string(rune(maxLength/100+'0')) + string(rune((maxLength/10)%10+'0')) + string(rune(maxLength%10+'0')) + " characters"
	}

	// Check for at least one alphanumeric character
	hasAlphaNum := false
	for _, r := range name {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			hasAlphaNum = true
			break
		}
	}
	if !hasAlphaNum {
		return fieldName + " must contain at least one letter or number"
	}

	return ""
}

// SanitizeAndValidate sanitizes input and validates it
func SanitizeAndValidate(input string, fieldName string, minLength int, maxLength int) (string, string) {
	sanitized := Name(input, maxLength)
	err := ValidateName(sanitized, fieldName, minLength, maxLength)
	return sanitized, err
}
